spring-xd-dirt/src/main/java/org/springframework/xd/dirt/stream/XDParser.java

```diff
@@ -43,4 +45,29 @@
 	 *         described by the DSL.
 	 */
 	List<ModuleDescriptor> parse(String name, String config, ParsingContext type);
+
+	public static class DefinitionOrException {
```

Comment:
I won't try to argue with you, especially in written form, I know where this leads :)

I like the decorated approach.

About the separation of syntactic / semantic split, please note that we already have two levels of parsers. BUT, the inner one had to know too much (mostly because of our previous approach to composed modules) and I would argue that the inner layer should not know about either the repository of existing streams (it does currently) nor things like module options (it does not atm). Those should be responsibilities of the outer layer.

For the record, I also think that the limitation we have where a stream can't be named after one of its module is now void and could be lifted.